src/lib/ptree/json.lua

```diff
@@ -135,18 +208,50 @@ local function write_json_string(output, str)
    output:write_chars('"')
 end
 
-function write_json_object(output, obj)
+write_json_type = function(output, var)
+   local tp = type(var)
+   if tp == 'table' then
+      if var == NIL then
+         return write_json_null(output)
+      end
+
+      for k, v in pairs(var) do
+         if not (type(k) == 'number' and math.floor(k)==k and 1<=k) then
+            return write_json_object(output, var)
+         end
+      end
+      return write_json_array(output, var)
+   elseif tp == 'string' then
+      return write_json_string(output, var)
+   elseif tp == 'boolean' or tp == 'number' then
+      return write_json_scalar(output, var)
+   end
+
+   return error('unable to serialize value of unknown type: '..tp)
+end
+
+write_json_object = function(output, obj)
    output:write_chars('{')
    local comma = false
    for k,v in pairs(obj) do
       if comma then output:write_chars(',') else comma = true end
-      write_json_string(output, k)
+      write_json_string(output, k) -- JSON keys must be strings
       output:write_chars(':')
-      write_json_string(output, v)
+      write_json_type(output, v)
    end
    output:write_chars('}')
 end
 
+write_json_array = function(output, obj)
+   output:write_chars('[')
+   local comma = false
+   for _,v in ipairs(obj) do
+      if comma then output:write_chars(',') else comma = true end
```

Comment:
OK, so what the `comma` flag does is to print out a ',' after the first iteration. What about:

```lua
for i,v in ipairs(obj) do
   if i > 1 then output:write_chars(',') end
```